pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/typedefinition/JavaTypeDefinition.java

```diff
@@ -52,33 +51,22 @@ public static JavaTypeDefinition forClass(TypeDefinitionType type, JavaTypeDefin
         }
     }
 
+    public static JavaTypeDefinition forClass(final Class<?> clazz) {
+        return clazz == Object.class ? JavaTypeDefinitionSimple.OBJECT_DEFINITION
+                                     : forClass(clazz, NO_GENERICS); // very common
+
+    }
+
     public static JavaTypeDefinition forClass(final Class<?> clazz, JavaTypeDefinition... boundGenerics) {
         if (clazz == null) {
             return null;
         }
 
-        // deal with generic types
-        if (boundGenerics.length != 0) {
-            // With generics there is no cache
-            return new JavaTypeDefinitionSimple(clazz, boundGenerics);
-        }
-
-        final JavaTypeDefinition typeDef = CLASS_EXACT_TYPE_DEF_CACHE.get(clazz);
-
-        if (typeDef != null) {
-            return typeDef;
-        }
-
-        final JavaTypeDefinition newDef;
         try {
-            newDef = new JavaTypeDefinitionSimple(clazz);
-        } catch (final NoClassDefFoundError e) {
+            return new JavaTypeDefinitionSimple(clazz, boundGenerics);
+        } catch (final LinkageError e) {
```

Comment:
I think, we should also catch here TypeNotPresentException, see also #2940

Reply:
Hm... maybe not - we don't seem to load anything here anymore. So I think, not even LinkageError would occur here... but somewhere else....